src/base58.rs

```diff
@@ -0,0 +1,441 @@
+                    continue;
+                }
+                encoding_flag = true;
+            }
+
+            let mut carry = b58_value as u32;
+            let mut i = 0;
+
+            while carry != 0 || i < size {
+                if first_byte_iter {
+                    raw_bytes.push(0);
+                    first_byte_iter = false;
+                } else {
+                    if let Some(raw_byte) = raw_bytes.get(i) {
+                        carry += *raw_byte as u32 * 58;
+                    } else {
+                        raw_bytes.push(0);
+                    }
+                }
+
+                raw_bytes[i] = (carry % 256) as u8;
+                carry /= 256;
+
+                i += 1
+            }
+
+            size = i;
+        }
+
+        raw_bytes.reverse();
+        final_bytes.extend_from_slice(&raw_bytes);
+        Ok(final_bytes)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum FromBase58CheckError {
+    Base58Error(FromBase58Error),
+    InvalidCheckSum,
+}
+
+#[derive(Debug, PartialEq)]
+enum Base58CheckVersion {
+    PKH,
+    SH,
+    SKWIF,
+    SKBIP38,
+    PKBIP32,
+}
+
+impl Base58CheckVersion {
+    fn to_vec(&self) -> Vec<u8> {
+        match self {
+            Base58CheckVersion::PKH => vec![0x00],
+            Base58CheckVersion::SH => vec![0x05],
+            Base58CheckVersion::SKWIF => vec![0x80],
+            Base58CheckVersion::SKBIP38 => vec![0x01, 0x42],
+            Base58CheckVersion::PKBIP32 => vec![0x04, 0x88, 0xB2, 0x1E],
+        }
+    }
+
+    fn get_version(payload_with_version: &[u8]) -> Option<Self> {
+        match payload_with_version {
+            [0x00, ..] => Some(Base58CheckVersion::PKH),
+            [0x05, ..] => Some(Base58CheckVersion::SH),
+            [0x80, ..] => Some(Base58CheckVersion::SKWIF),
+            [0x01, 0x42, ..] => Some(Base58CheckVersion::SKBIP38),
+            [0x04, 0x88, 0xB2, 0x1E, ..] => Some(Base58CheckVersion::PKBIP32),
+            _ => None,
+        }
+    }
+}
+
+trait ToBase58Check {
+    fn to_base58check(&self, version: Option<Base58CheckVersion>) -> String;
+}
+
+trait FromBase58Check {
+    fn from_base58check(
+        &self,
+    ) -> Result<(Vec<u8>, Option<Base58CheckVersion>), FromBase58CheckError>;
+}
+
+impl ToBase58Check for [u8] {
+    fn to_base58check(&self, version: Option<Base58CheckVersion>) -> String {
+        let mut final_vec = match version {
+            None => vec![],
+            Some(v) => v.to_vec(),
+        };
+
+        final_vec.extend_from_slice(self);
+
+        let final_hash = double_sha256(&final_vec);
+        final_vec.extend_from_slice(&final_hash[0..4]);
+
+        let base58_encoded = final_vec.as_slice().to_base58();
+        base58_encoded
+    }
+}
+
+impl FromBase58Check for str {
+    fn from_base58check(
+        &self,
+    ) -> Result<(Vec<u8>, Option<Base58CheckVersion>), FromBase58CheckError> {
+        let decoded = match self.from_base58() {
```

Comment:
Style opinion:

Here you can remove the match and use the `map_err` method instead, since you're doing no transformation in the `Ok` case, like this:

```rust
let decoded = self.from_base58().map_err(FromBase58CheckError::Base58Error)?;
```

Reply:
Oh that's perfect! I will implement your suggestion :)